app/celery/process_delivery_status_result_tasks.py

```diff
@@ -1,10 +1,295 @@
+        check_and_queue_callback_task(notification, payload)
+        return True
+
+    except Retry:
+        # This block exists to preempt executing the "Exception" logic below.  A better approach is
+        # to catch specific exceptions where they might occur.
+        raise
+    except Exception as e:
+        current_app.logger.exception(e)
+        self.retry(queue=QueueNames.RETRY)
+
     return True
+
+
+def attempt_to_get_notification(
+        reference: str, notification_status: str, event_timestamp_in_ms: str
+) -> Tuple[Notification, bool, bool]:
+    """ Attempt to get the notification object and determine whether the Celery Event should be retry or exit"""
+    should_retry = False
+    notification = None
+    try:
+        notification = dao_get_notification_by_reference(reference)
+        should_exit = check_notification_status(notification, notification_status)
+    except NoResultFound:
+        message_time = datetime.datetime.fromtimestamp(
+            math.floor(float(event_timestamp_in_ms) / 1000)
+        )
+        if datetime.datetime.utcnow() - message_time < datetime.timedelta(minutes=5):
+            current_app.logger.info(
+                "Delivery Status callback event for reference %s was received less than five minutes ago.",
+                reference,
+            )
+            should_retry = True
+        else:
+            current_app.logger.warning(
+                "notification not found for reference: %s (update to %s)",
+                reference,
+                notification_status,
+            )
+        statsd_client.incr("callback.delivery_status.no_notification_found")
+        should_exit = True
+    except MultipleResultsFound:
+        current_app.logger.warning(
+            "multiple notifications found for reference: %s (update to %s)",
+            reference,
+            notification_status,
+        )
+        statsd_client.incr("callback.delivery_status.multiple_notifications_found")
+        should_exit = True
+
+    return notification, should_retry, should_exit
+
+
+def log_notification_status_warning(notification, status: str) -> None:
+    time_diff = datetime.datetime.utcnow() - (notification.updated_at or notification.created_at)
+    current_app.logger.warning(
+        "Invalid callback received. Notification id %s received a status update to %s "
+        "%s after being set to %s. %s sent by %s",
+        notification.id,
+        status,
+        time_diff,
+        notification.status,
+        notification.notification_type,
+        notification.sent_by,
+    )
+
+
+def check_notification_status(notification: Notification, notification_status: str) -> bool:
+    # Do not update if the status has not changed.
+    if notification_status == notification.status:
+        current_app.logger.info(
+            "SQS callback received the same status of %s for notification %s)",
+            notification_status,
+            notification_status,
```

Comment:
This is printing the same thing twice, the second `notification_status`, is probably supposed to be `notification.id`. There's also a random end parenthesis `)` in the string, maybe remove that?